src/main/java/com/mongodb/hibernate/internal/type/ValueConversions.java

```diff
@@ -184,17 +185,19 @@ private static BsonArray arrayToBsonValue(Object value) throws SQLFeatureNotSupp
         } else if (value instanceof BsonDecimal128 v) {
             return toDomainValue(v);
         } else if (value instanceof BsonString v) {
-            return toDomainValue(v, domainType);
+            return uncheckedToDomainValue(v, domainType);
         } else if (value instanceof BsonBinary v) {
             return toDomainValue(v);
         } else if (value instanceof BsonObjectId v) {
             return toDomainValue(v);
         } else if (value instanceof BsonArray v && domainType.isArray()) {
             return toDomainValue(v, assertNotNull(domainType.getComponentType()));
         }
-        throw new SQLFeatureNotSupportedException(format(
-                "Value [%s] of type [%s] is not supported for the domain type [%s]",
-                value, assertNotNull(value).getClass().getTypeName(), domainType));
+        throw exceptionDomainTypeUnsupportedOrMustBeExplicit(value, domainType);
+    }
+
+    public static @Nullable Object toDomainValue(BsonValue value) throws SQLFeatureNotSupportedException {
```

Comment:
This PR only needed this new method. However, I also wanted to stop using `Object.class` to represent "unknown domain type", which caused the rest of the changes.